iceoryx_utils/include/iceoryx_utils/internal/cxx/vector.inl

```diff
@@ -259,66 +271,77 @@ const T& vector<T, Capacity>::front() const noexcept
 {
     if (empty())
     {
         std::cerr << "Attempting to access the back of an empty vector!" << std::endl;
         std::terminate();
     }
-    return at(size() - 1);
+    return at(this->size() - 1);
 }
 
 template <typename T, uint64_t Capacity>
-inline typename vector<T, Capacity>::iterator vector<T, Capacity>::begin()
+inline typename vector<T, Capacity>::iterator vector<T, Capacity>::begin() noexcept
 {
-    return reinterpret_cast<iterator>(&(m_data[0]));
+    return reinterpret_cast<iterator>(this->data());
 }
 
 template <typename T, uint64_t Capacity>
-inline typename vector<T, Capacity>::const_iterator vector<T, Capacity>::begin() const
+inline typename vector<T, Capacity>::const_iterator vector<T, Capacity>::begin() const noexcept
 {
-    return reinterpret_cast<const_iterator>(&(m_data[0]));
+    return reinterpret_cast<const_iterator>(this->data());
 }
 
 template <typename T, uint64_t Capacity>
-inline typename vector<T, Capacity>::iterator vector<T, Capacity>::end()
+inline typename vector<T, Capacity>::iterator vector<T, Capacity>::end() noexcept
 {
-    return reinterpret_cast<iterator>((&(m_data[0]) + m_size)[0]);
+    return this->data() + this->size();
 }
 
 template <typename T, uint64_t Capacity>
-inline typename vector<T, Capacity>::const_iterator vector<T, Capacity>::end() const
+inline typename vector<T, Capacity>::const_iterator vector<T, Capacity>::end() const noexcept
 {
-    return reinterpret_cast<const_iterator>((&(m_data[0]) + m_size)[0]);
+    return reinterpret_cast<const_iterator>(this->data() + this->size());
 }
 
 template <typename T, uint64_t Capacity>
-inline typename vector<T, Capacity>::iterator vector<T, Capacity>::erase(iterator position)
+inline typename vector<T, Capacity>::iterator vector<T, Capacity>::erase(const_iterator position)
 {
-    if (begin() <= position && position < end())
+    /// If position < begin(), this shall return end(), which is guaranteed:
+    /// WLOG, let vector v have Capacity=(2^64-begin())/sizeof(T)-sizeof(m_size)
+    /// (note that iterators, index, and Capacity are all limited by 64 bit address space)
+    /// With position=0, position-begin()==(2^64-begin())/sizeof(T) > Capacity => returns end()
+    uint64_t index = static_cast<uint64_t>(position - begin());
+    return erase(index);
+}
+
+template <typename T, uint64_t Capacity>
+inline typename vector<T, Capacity>::iterator vector<T, Capacity>::erase(uint64_t index)
+{
+    uint64_t n = index;
+    if (n >= this->size())
     {
-        uint64_t index = static_cast<uint64_t>(position - begin()) % (sizeof(element_t) * Capacity);
-        size_t n = index;
-        for (; n + 1u < size(); ++n)
-        {
-            at(n) = std::move(at(n + 1u));
-        }
-        at(n).~T();
-        m_size--;
+        return end();
+    }
+    at(index).~T();
+    for (; n + 1u < this->size(); ++n)
+    {
+        new (&at(n)) T(std::move(at(n + 1u)));
```

Comment:
Here you would have undefined behavior. Let's assume we have a vector with the position[content] { 1[a], 2[b], 3[c], 4[d], 5[e] }.

We call again
```cpp
myVector.erase(2);
```

1. DTor of position 3 is called
2. Placement new at position 3 with move ctor of content [d] of position 4.
3. Placement new at position 4 with move ctor of content [e] of position 5. This is undefined behavior since you override the memory of an existing object without calling the dtor before.

Reply:
@elfenpiff I updated the PR and now also call the dtor of the moved objects.